control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -847,6 +867,39 @@ func (r *HostedControlPlaneReconciler) validateConfigAndClusterCapabilities(hc *
 	return nil
 }
 
+func (r *HostedControlPlaneReconciler) validateHostedClusterConfiguration(ctx context.Context, hc *hyperv1.HostedControlPlane, guestNodesGetter func(context.Context, *hyperv1.HostedControlPlane) (corev1client.NodesGetter, error)) error {
+	if hc.Spec.Configuration == nil || hc.Spec.Configuration.Node == nil {
+		return nil
+	}
+	if minVersion := hc.Spec.Configuration.Node.MinimumKubeletVersion; minVersion != "" {
+		nodesClient, err := guestNodesGetter(ctx, hc)
+		if err != nil {
+			// TODO: should we not error here?
+			// If we fail to get the guest client then maybe the guest
+			// cluster isn't setup yet, in which case we don't need to do any
+			// validation that all of the existing nodes are new enough
+			return fmt.Errorf("failed to get guest client %v", err)
+		}
+		if err := nodelib.ValidateMinimumKubeletVersion(nodesClient, minVersion); err != nil {
+			return fmt.Errorf("validating failed for %s: %v", minVersion, err)
+		}
+	}
+	return nil
+}
+
+// A helper function to factor out the pieces that rely on a functioning guest cluster, added for unit testing purposes.
+func (r *HostedControlPlaneReconciler) getHostedClusterNodesClient(ctx context.Context, hc *hyperv1.HostedControlPlane) (corev1client.NodesGetter, error) {
```

Comment:
hypershift cohesively groups adhoc controllers that need to reconcile guest cluster resources within the hostedclusterconfigoperator.
See https://github.com/openshift/hypershift/tree/main/control-plane-operator/hostedclusterconfigoperator/controllers/nodecount

Please add the logic to set the current minimum version in the HCP there (similar to what we are doing for node count). Ideally this logic would its own controller that is instantiated here https://github.com/openshift/hypershift/blob/main/control-plane-operator/hostedclusterconfigoperator/controllers/nodecount/setup.go#L22 so then we can run it only when the MinimumKubeletVersion gate is enabled

Reply:
are you opposed to me running it in https://github.com/openshift/hypershift/blob/main/control-plane-operator/hostedclusterconfigoperator/controllers/node/node.go instead and gating that behavior on the feature gate, to reduce bolierplate?

Reply:
alrighty we chatted offline and the new controller was suggested, so I've adopted that and used that to pass the oldestKubeletVersion to the HCP controller